### PR TITLE
Introduce FP32 and FP16 zero_point for CPU

### LINKED ISSUE #61866: [quant] Add support for floating point zero_point in fake_quantize ops
Currently `fake_quantize_per_tensor/per_channel` operators and kernels expect int32 zero_point values.
For some operations we require to use floating point zero_point values to reduce quantization error. 

Example of where this is currently implemented is in `quantize_per_channel` and corresponding implementation of quantize function is in https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/affine_quantizer_base.cpp#L169

Specifically this issue tracks support for 
1. Fp32 scale/zero_point support in fake_quantize_per_channel for 8 bit embedding ops
2. FP16 scale/zero_point in fake_quantize_per_channel/quantize_per_channel for 4-bit embedding bag ops.

Required for Embedding/EmbeddingBag QAT support #61865

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo

### PR DESCRIPTION
**Pull request by b-koopman**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64969
* __->__ #63996

Differential Revision: [D30918012](https://our.internmc.facebook.com/intern/diff/D30918012)

### CODE REVIEW

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30918012).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30918012).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30918012).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30918012).

**Comment by b-koopman** [comment]
@b-koopman has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30918012).

File: test/typing/pass/math_ops.py [hunk 1]
```diff
@@ -122,9 +122,17 @@
...
-zero_points = torch.zeros(2).to(torch.long)
+zero_points = torch.zeros(2).to(torch.int)
+torch.fake_quantize_per_channel_affine(x, scales, zero_points, 1, 0, 255)
+
+# fake_quantize_per_channel_affine float zero_point
+zero_points = torch.zeros(2).to(torch.float)
+torch.fake_quantize_per_channel_affine(x, scales, zero_points, 1, 0, 255)
+
+# fake_quantize_per_channel_affine fp16 zero_point
+zero_points = torch.zeros(2).to(torch.half)
```
[the remark this answers is not in the document]
Comment (supriyar) [reply test/typing/pass/math_ops.py:135]: @b-koopman  can we revert changes to this file?

**Comment by supriyar** [comment]
The test failures appear to be related. Do we need to modify the reference implementation in the tests as well to account for the floating point zp?

File: test/typing/pass/math_ops.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (b-koopman) [reply test/typing/pass/math_ops.py]: Done, I noticed a lot of these tests fail, is this basically a deprecated test?

**Pull request closed without merging**